homeassistant/components/netgear/router.py

```diff
@@ -0,0 +1,125 @@
+"""Represent the Netgear router and its devices."""
+from datetime import timedelta
+from typing import Dict
+
+from pynetgear import Netgear
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_DEVICES,
+    CONF_EXCLUDE,
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_SSL,
+    CONF_URL,
+    CONF_USERNAME,
+)
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import HomeAssistantType
+
+from .const import DOMAIN
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+
+class NetgearRouter:
+    """Representation of a Netgear router."""
+
+    def __init__(self, hass: HomeAssistantType, entry: ConfigEntry) -> None:
+        """Initialize a Netgear router."""
+        self.hass = hass
+        self._url = entry.data.get(CONF_URL)
+        self._host = entry.data.get(CONF_HOST)
+        self._port = entry.data.get(CONF_PORT)
+        self._ssl = entry.data.get(CONF_SSL)
+        self._username = entry.data.get(CONF_USERNAME)
+        self._password = entry.data[CONF_PASSWORD]
+        self._tracked_devices = entry.data.get(CONF_DEVICES)
+        self._excluded_devices = entry.data.get(CONF_EXCLUDE)
+
+        self._api: Netgear = None
+        self._attrs = {}
+
+        self.devices: Dict[str, any] = {}
+
+        self._unsub_dispatcher = None
+        self.listeners = []
+
+    async def setup(self) -> None:
+        """Set up a Netgear router."""
+        self._api = await self.hass.async_add_executor_job(
+            Netgear,
+            self._password,
+            self._host,
+            self._username,
+            self._port,
+            self._ssl,
+            self._url,
+        )
+
+        await self.hass.async_add_executor_job(self._api.login)
+
+        await self.update_devices()
+        self._unsub_dispatcher = async_track_time_interval(
+            self.hass, self.update_devices, SCAN_INTERVAL
+        )
+
+    def unload(self) -> None:
```

Comment:
It helps our understanding of teh code if functions that are supposed to be run in teh event loop (async or `@callback`) have their name start with `async_`. You can import the `@callback` from `homeassistant.core` and decorate functions that are ok for the event loop.